### PR TITLE
Fix issue Phing targets are not executed in the build dir

### PR DESCRIPTION
**Pull request by corretge**
See comments in the issues #746 and #748 and commits 56f04d9 and ed13ec5.

Other commits without linked official issue has been attached to the pull request. Maybe could be interesting for the project:

* #746 fix Field 'status' doesn't have a default value when click "Rebuild Now"
* #748 fix Phing targets are not executed in the build dir
* Fix issue build_settings ignored in createWorkingCopy.
* Add a new option in the **build** section: prefer_rsync that exclude the ignored folders in the copy when the project is hosted in **Local Path**.
* Flush stdout while executing a step, don't wait at the end of the execution.
* Improve to log more colours from console stdout than red and green.
* Allow to override phpci.yml in a concrete build.
* Fix phpunit bootstrap that crash if not exists a config file.

### CODE REVIEW

**Comment by corretge** [comment]
Can anybody show me the details of the failure?

**Comment by corretge** [comment]
Wow! the only ones commits for this pull request are the first two commits.

**Comment by corretge** [comment]
Can a PHPCI collaborator attach information about the failed build? thanks

**Comment by tvbeek** [comment]
@dancryer can do that

**Comment by corretge** [comment]
Thanks @tvbeek and @dancryer, @wandersonwhcr  explain me how to reproduce the build process in local. I had errors whe try to pass phpcs and phpms checkings.

**Comment by wandersonwhcr** [comment]
:smiley: you're welcome!

**Comment by wandersonwhcr** [comment]
@corretge Thanks! I need the PHPUnit language config fix to upload my ApiGen plugin.

**Comment by corretge** [comment]
@wandersonwhcr thanks to you, I will use your ApiGen plugin when merged. I'm a phpdoc user but ApiGen seems greater.

File: PHPCI/Controller/BuildController.php [hunk 1]
```diff
@@ -179,15 +179,49 @@ public function delete($buildId)
...
 
     /**
-    * Parse log for unix colours and replace with HTML.
-    */
+     * Parse log for unix colours and replace with HTML.
+     *
+     * @link http://tldp.org/HOWTO/Bash-Prompt-HOWTO/x329.html
+     * @param string $log Output
+     */
     protected function cleanLog($log)
```
Comment (tvbeek) [inline PHPCI/Controller/BuildController.php]: Can you combine all the str_replace in one call? Now you have a str_replace with an array and after that two calls with one value. Or can you add comments if that is needed

File: PHPCI/Helper/BaseCommandExecutor.php [hunk 2]
```diff
@@ -81,7 +81,14 @@ public function executeCommand($args = array())
 
         if ($this->quiet) {
             $this->logger->log('Executing: ' . $command);
+        } else {
+            /**
+             * show the command to execute in purple
+             * @link http://tldp.org/HOWTO/Bash-Prompt-HOWTO/x329.html
+             */
+            $this->logger->log([null, null, "\033[0;35mExecuting: {$command}\033[0m", null]);
```
Comment (tvbeek) [inline PHPCI/Helper/BaseCommandExecutor.php]: We normally don't place the vars in the string, can you change it to:
$this->logger->log([null, null, "\033[0;35mExecuting: " . $command . "\033[0m", null]);

File: PHPCI/Helper/BaseCommandExecutor.php [hunk 3]
```diff
@@ -92,30 +99,25 @@ public function executeCommand($args = array())
...
-        if ($shouldOutput && !empty($this->lastOutput)) {
+        /**
+         * If we don't output passthru but we have an error, log it.
+         */
+        if (!$shouldOutput && $status != 0 && !empty($this->lastOutput)) {
             $this->logger->log($this->lastOutput);
         }
 
         if (!empty($this->lastError)) {
-            $this->logger->log("\033[0;31m" . $this->lastError . "\033[0m", LogLevel::ERROR);
```
Comment (tvbeek) [inline PHPCI/Helper/BaseCommandExecutor.php]: Can you explain why you change this line? Maybe I miss the reason

File: PHPCI/Helper/BaseCommandExecutor.php [hunk 4]
```diff
@@ -128,6 +130,66 @@ public function executeCommand($args = array())
     }
 
     /**
+     * Manage the standard output of a process execution flushing data in
+     * realtime
+     *
```
Comment (tvbeek) [inline PHPCI/Helper/BaseCommandExecutor.php]: Can you add the process parameter to the doc block?

File: PHPCI/Helper/BaseCommandExecutor.php [hunk 5]
```diff
@@ -128,6 +130,66 @@ public function executeCommand($args = array())
...
+     */
+    protected function manageExecutionStandardOutput($process, $pipes, $shouldOutput)
+    {
+        $status = false;
+
+        if (is_resource($process)) {
+            /**
+             * don't wait to the end of the process to output stdout
+             */
+            while (!feof($pipes[1])) {
```
Comment (tvbeek) [inline PHPCI/Helper/BaseCommandExecutor.php]: are you sure that $pipes[1] is always available? Else you need to check it (maybe in the if on line 143)

File: PHPCI/Controller/BuildController.php (hunk 1 above)
Comment (corretge) [inline PHPCI/Controller/BuildController.php]: Yes of course! I leave the oldest calls by mistake!

File: PHPCI/Helper/BaseCommandExecutor.php (hunk 4 above)
Comment (corretge) [inline PHPCI/Helper/BaseCommandExecutor.php]: done!

**Comment by dancryer** [comment]
There are way too many changes in this PR for us to merge it as one thing. 

Please can you split this into logical separate pull requests. For example the formatting of the log should be a PR on its own, same for changing how stdout is handled, etc.

**Comment by corretge** [comment]
Ok, I will split it per feature.

**Pull request closed without merging**